### PR TITLE
Multi language donation items

### PR DESCRIPTION
**Pull request by ManuelFeller**
Multi-Language CSV Data Source plus adoption for DonationRow and DonationPill (new CSV URL is in the Wiki of the private repo with project specifics)

### CODE REVIEW

**Comment by ManuelFeller** [comment]
sorry for the large `dataStore.ts` diff, have run prettier :(

File: donation-tracker/src/components/donationPill.tsx [hunk 1]
```diff
@@ -1,16 +1,16 @@
...
+import { useTranslation, I18nextContext } from 'gatsby-plugin-react-i18next';
 import { Button } from '@mui/material';
 
 import { DonationItem } from '../types/donationItem';
 import { Campaign } from '../types/campaign';
-import { useTranslation } from 'react-i18next';
 
 const DonationPill = ({ campaign, donationItem }: { donationItem: DonationItem; campaign: Campaign }) => {
   const { t } = useTranslation();
+  const langContext = React.useContext(I18nextContext);
```
Comment (cschweikert) [inline donation-tracker/src/components/donationPill.tsx:10]: ah, good to know. I'll also need this soon

File: donation-tracker/src/utils/dataStore.ts [hunk 2]
```diff
@@ -1,316 +1,486 @@
...
+    this.debugLog('DataStore: Creating instance, initializing internal variables');
+    this.updateSubscribers = new Array<Function>();
+    this.isRefreshing = false;
+    this.localData = null;
+    this.hasLocalData = false;
+    this.debugLog('DataStore: trying to read local storage data');
+    let tmpData = null;
+    if (typeof localStorage === 'undefined') {
+      this.isGatsbyBuild = true;
+      tmpData = `{"version":"${this.dataVersion}","timeStamp":"2025-01-01T01:23:45.678Z","requestTime":"2025-01-01T01:23:45.678Z","data":[{"article":{"en":"Backpack 50-70 lt","de":"Rucksack 50-70 l"},"campaignKey":"demo","neededOverall":10,"alreadyDonated":6,"remainingNeed":4,"unit":{"en":"pcs","de":"Stk"},"formLink":{"en":"https://google.com","de":"https://google.de"}}]}`;
```
Comment (cschweikert) [inline donation-tracker/src/utils/dataStore.ts]: might look much nicer to set it like
```
tmpData = JSON.stringyfy({
  version: this.dataVersion,
  timeStamp: ...
  ...
}
```

File: donation-tracker/src/utils/dataStore.ts [hunk 3]
```diff
@@ -1,316 +1,486 @@
...
-	 * @returns A boolean information if a refresh trigger will execute or not
-	 */
-	public canRefreshAlready() {
-		return this.isDataOutdated();
-	}
+  /**
+   * Function to check in how many seconds the next data refresh can / will be performed
+   * @returns The number of seconds until the next data refresh - or undefined if it can not be calculated
+   */
+  public getSecondsUntilRefresh(): number | undefined {
```
Comment (cschweikert) [inline donation-tracker/src/utils/dataStore.ts:154]: or shorter like this:
```
  public getSecondsUntilRefresh(): number | undefined {
    if (this.isRefreshing) {
      return 0;
    }
    if (this.localData === null) {
      return undefined;
    }
    const waitingTime = 1000 * 60 * this.maxDataAgeInMinutes;
    const nextUpdateTime = new Date(this.localData.requestTime).getTime() + waitingTime;
    return Math.max(0, Math.round((nextUpdateTime - Date.now()) / 1000));
  }
```
🤓🤓🤓

File: donation-tracker/src/utils/dataStore.ts [hunk 4]
```diff
@@ -1,316 +1,486 @@
...
-				tmpResult.push(tmpObject);
-			}
-		}
-		return tmpResult;
-	}
+  /**
+   * Internal function to check if the expected data version of the code matches the data version found locally
+   * @returns true if version matches with code, false if not
+   */
+  private isLocalDataVersionOk() {
```
Comment (cschweikert) [inline donation-tracker/src/utils/dataStore.ts:201]: short
```
  private isLocalDataVersionOk() {
    this.debugLog('DataStore: Data Version Check');
    // version field is available and matches current version
    return this.localData?.version !== undefined && this.localData?.version === this.dataVersion;
  }
```

or maybe shorter

```
  private isLocalDataVersionOk() {
    this.debugLog('DataStore: Data Version Check');
    // version field matches current version
    return this.localData?.version === this.dataVersion;
  }
```

File: donation-tracker/src/utils/dataStore.ts [hunk 5]
```diff
@@ -1,316 +1,486 @@
...
+    this.hasLocalData = false;
+    if (!this.isGatsbyBuild) setTimeout(this.refreshData.bind(this), 100);
+  }
+
+  /**
+   * Internal function to register the next data update
+   */
+  private registerDataUpdate() {
+    if (PageConfiguration.AutoRefresh) {
+      let secondsUntilRefresh = this.getSecondsUntilRefresh();
```
Comment (cschweikert) [inline donation-tracker/src/utils/dataStore.ts]:
```
    // use default refresh time, if it can not be calculated
    const secondsUntilRefresh = this.getSecondsUntilRefresh() ?? 1000 * 60 * this.maxDataAgeInMinutes;
```

File: donation-tracker/src/utils/dataStore.ts [hunk 6]
```diff
@@ -1,316 +1,486 @@
...
+  /**
+   * Internal function to map the columns to the data
+   * @param csvConfigRow The first row of the CVS file / export
+   * @returns A ColumnMapping object
+   */
+  private createColumnMapping(csvConfigRow: string): ColumnMapping {
+    this.debugLog('DataStore: Creating column mapping');
+    const headers = this.parseCsvLine(csvConfigRow.toLowerCase());
+    const result: ColumnMapping = {
+      languages: new Array<string>(),
```
Comment (cschweikert) [inline donation-tracker/src/utils/dataStore.ts:239]:
```suggestion
      languages: [],
```

File: donation-tracker/src/utils/dataStore.ts [hunk 7]
```diff
@@ -1,316 +1,486 @@
...
+      need: -1,
+      donated: -1,
+      lastChange: -1,
+    };
+    headers.forEach((columnContent, index) => {
+      if (columnContent.toString().indexOf(':') > -1) {
+        // we have a localized value here
+        const pair = columnContent.toString().split(':');
+        const langCode = pair[1].toString().trim();
+        let langIndex = result.languages.indexOf(langCode);
```
Comment (cschweikert) [inline donation-tracker/src/utils/dataStore.ts:253]: alternative:
```
        if (!result.languages.includes(langCode)) {
          result.languages.push(langCode);
        }
```

File: donation-tracker/src/utils/dataStore.ts [hunk 8]
```diff
@@ -1,316 +1,486 @@
...
+      item: {},
+      unit: {},
+      form: {},
+      campaignId: -1,
+      need: -1,
+      donated: -1,
+      lastChange: -1,
+    };
+    headers.forEach((columnContent, index) => {
+      if (columnContent.toString().indexOf(':') > -1) {
```
Comment (cschweikert) [inline donation-tracker/src/utils/dataStore.ts:249]:
```suggestion
      if (columnContent.toString().includes(':')) {
```

File: donation-tracker/src/utils/dataStore.ts [hunk 9]
```diff
@@ -1,316 +1,486 @@
...
+          case 'need':
+            result.need = index;
+          case 'donated':
+            result.donated = index;
+          case 'lastchange':
+            result.lastChange = index;
+        }
+      }
+    });
+    return result;
```
Comment (cschweikert) [inline donation-tracker/src/utils/dataStore.ts:279]: I wonder, if we can not reshape the whole function in the following way (rough sketch):
```
const headers = this.parseCsvLine(csvConfigRow.toLowerCase()).map(columnContent => columnContent.toString().trim());
return = {
      languages: uniq(headers.filter(header => header.includes(':'))) // or similar to maybe not use lodash
      item: headers.filter(header => header.includes('item:')).reduce((result, header, index)=>({...result, [getLangCode(header)]: index}), {})
      unit: ...,
      form: ...,
      campaignId: headers.indexOf(header => header === 'campaign'),
      need: headers.indexOf(header => header === 'need'),
      donated: headers.indexOf(header => header === 'donated'),
      lastChange: headers.indexOf(header => header === 'lastchange'),
    };
```

File: donation-tracker/src/utils/dataStore.ts [hunk 10]
```diff
@@ -1,316 +1,486 @@
...
+  }
+
+  /**
+   * Internal function to load / refresh the local data
+   */
+  private async refreshData() {
+    this.isRefreshing = true;
+    this.debugLog('DataStore: Data refresh started');
+    let data = await fetch(this.dataSourceUrl, { cache: 'no-store' });
```
Comment (cschweikert) [inline donation-tracker/src/utils/dataStore.ts]:
```suggestion
    const data = await fetch(this.dataSourceUrl, { cache: 'no-store' });
```
??

File: donation-tracker/src/utils/dataStore.ts [hunk 11]
```diff
@@ -1,316 +1,486 @@
...
+      const mapping = this.createColumnMapping(rows[0].trim());
+      if (!this.validateColumnMapping(mapping)) {
+        this.isRefreshing = false;
+        this.registerDataUpdate();
+        return;
+      }
+
+      this.debugLog('DataStore: Data received');
+      // extract line with change timestamp
+      let timestampRow = this.parseCsvLine(rows[1].trim());
```
Comment (cschweikert) [inline donation-tracker/src/utils/dataStore.ts]:
```suggestion
      const timestampRow = this.parseCsvLine(rows[1].trim());
```
??

File: donation-tracker/src/utils/dataStore.ts [hunk 12]
```diff
@@ -1,316 +1,486 @@
...
+    const minColumns = 3 + mapping.languages.length * 3;
+    for (let idx = 2; idx < dataRows.length; idx++) {
+      const tmpLineData = this.parseCsvLine(dataRows[idx].trim());
+      if (tmpLineData.length >= minColumns) {
+        const tmpObject: DonationItem = {
+          article: {},
+          campaignKey: tmpLineData[mapping.campaignId].toString(),
+          neededOverall: Number.parseInt(tmpLineData[mapping.need].toString()),
+          alreadyDonated: Number.parseInt(tmpLineData[mapping.donated].toString()),
+          remainingNeed: Number.parseInt(tmpLineData[mapping.need].toString()) - Number.parseInt(tmpLineData[mapping.donated].toString()),
```
Comment (cschweikert) [inline donation-tracker/src/utils/dataStore.ts]:
```suggestion
          remainingNeed: Math.max(0, Number.parseInt(tmpLineData[mapping.need].toString()) - Number.parseInt(tmpLineData[mapping.donated].toString())),
```
there might be cases of more donations than needed. Maybe the remaining need should then not be negative.

File: donation-tracker/src/utils/dataStore.ts [hunk 13]
```diff
@@ -1,316 +1,486 @@
...
-	 * Internal helper to print debug messages to teh console log (if activated in teh config)
-	 * @param message The message (or object) to log
-	 */
-	private debugLog(message: any) {
-		 if (PageConfiguration.LogToConsole) {
-			console.debug(message);
-		 }
-	}
+  /**
+   * Internal helper to print debug messages to the console log (if activated in the config)
```
Comment (cschweikert) [inline donation-tracker/src/utils/dataStore.ts:468]: 😁

File: donation-tracker/src/utils/dataStore.ts [hunk 14]
```diff
@@ -1,316 +1,486 @@
...
+      if (!this.validateColumnMapping(mapping)) {
+        this.isRefreshing = false;
+        this.registerDataUpdate();
+        return;
+      }
+
+      this.debugLog('DataStore: Data received');
+      // extract line with change timestamp
+      let timestampRow = this.parseCsvLine(rows[1].trim());
+      if (timestampRow.length >= mapping.lastChange + 1) {
```
Comment (cschweikert) [inline donation-tracker/src/utils/dataStore.ts:351]: I wonder if this check is still needed since we now have somewhat "absolute positioning" of columns by full header detection.

**Review by cschweikert — APPROVED** [review]
github didn't want to present me the diff for `dataStore` right away!?! Challenge accepted! 🤪

tl;dr
no real blocker, feel free to fix or merge

File: donation-tracker/src/components/donationPill.tsx (hunk 1 above)
Comment (ManuelFeller) [reply donation-tracker/src/components/donationPill.tsx:10]: Happy to be of service :D

File: donation-tracker/src/utils/dataStore.ts (hunk 2 above)
Comment (ManuelFeller) [reply donation-tracker/src/utils/dataStore.ts]: Will do

File: donation-tracker/src/utils/dataStore.ts (hunk 3 above)
Comment (ManuelFeller) [reply donation-tracker/src/utils/dataStore.ts:154]: good catch! Implemented

File: donation-tracker/src/utils/dataStore.ts (hunk 4 above)
Comment (ManuelFeller) [reply donation-tracker/src/utils/dataStore.ts:201]: I like this one too

File: donation-tracker/src/utils/dataStore.ts (hunk 5 above)
Comment (ManuelFeller) [reply donation-tracker/src/utils/dataStore.ts]: done

File: donation-tracker/src/utils/dataStore.ts (hunk 14 above)
Comment (ManuelFeller) [reply donation-tracker/src/utils/dataStore.ts:351]: I'll keep this as an empty column (someone accidentally deleting the date) would result in a "out of bounds" - as the RegEx row splitter can not really deal with empty columns (shortcoming to handle later; can be removed once this flaw is dealt with)
But valid point
